### PR TITLE
Paginate dataset image gallery

### PR DESCRIPTION
**Pull request by valerian**
# Feature

## Work performed

- Made the `ToggleButtonGroup` component generic, and changed its default selected background color from gray-200 to gray-300.
- On dataset images tab, made a working pagination interface at the bottom, according to #660 :
  - Page selector, similar to the image selector on the labelling view.
  - Segmented controls to select the number of items per page (50/200/1000).
  - Display the indexes of items displayed, such as "501-550 of 720".
  - Pagination tools are on a bottom bar, always visible (fixed position).

## Caveats

- The pagination toolbar gets covered by the cookie banner

## Resolved issues

Solves #660

### CODE REVIEW

**Comment by nicOwlas** [comment]
@valerian I have a 500 error when trying to sign-in. Are you having the same?

**Comment by nicOwlas** [comment]
- When I import images from the image gallery tab. The count of images is updated but not the cards in the gallery.
We should refresh the gallery to load the newly imported images (ideally refresh the gallery only when new elements were added to the page displayed to the user).
image

File: typescript/web/src/components/import-button/import-images-modal/import-images-modal.tsx [hunk 1]
```diff
@@ -51,6 +52,7 @@ export const ImportImagesModal = ({
         fetchPolicy: "network-only",
       });
       client.query({ query: getDatasetsQuery, fetchPolicy: "network-only" });
+      client.refetchQueries({ include: [paginatedImagesQuery] });
```
Comment (nacho-carnicero) [inline typescript/web/src/components/import-button/import-images-modal/import-images-modal.tsx:55]: This should be removed right?

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+          />
+        </Tooltip>
+      ) : (
+        <IconButton
+          disabled
+          bg="none"
+          aria-label="No previous page"
+          icon={<RiArrowLeftSLine size="1.5em" />}
+        />
+      )}
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-tool/pagination-tool.tsx:63]: Instead of doing a conditional rendering, I'd use the `isDisabled` prop of `Tooltip` and always render both components, disabling them both when `currentPage <=1`

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            setPage(inputValue);
+          }}
+        >
+          <NumberInputField
+            rounded={6}
+            onClick={(event) => {
+              (event.target as HTMLInputElement).select();
+            }}
+            textAlign="center"
+            w={`${digitCount * digitsPerRem + 1.5}rem`}
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-tool/pagination-tool.tsx:92]: I'd use here a Chakra value for the width, such as `2`, `3` or `4`. This enables to have a consistent design across the different components (same thing below)

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        mr={3}
+        ml={2}
+        w={`${digitCount * digitsPerRem + 0.5}rem`}
+        textAlign="left"
+        cursor="default"
+        fontSize="sm"
+      >
+        {maxPage}
+      </Text>
+      {currentPage < maxPage ? (
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-tool/pagination-tool.tsx:111]: Same thing as before, I'd avoid the conditional rendering by using the `isDisabled` prop of the Tooltip component

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+  useEffect(() => {
+    setInputValue(currentPage);
+  }, [currentPage]);
+
+  const handleKeyPress = (event: any) => {
+    if (event.key === "Enter") {
+      setPage(event.target.value as number);
+    }
+    if (event.key === "Escape" || event.key === "Esc") {
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-tool/pagination-tool.tsx:35]: This does not seem to work, at least in mac

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            if (!isNaN(valNum)) {
+              setInputValue(clamp(valNum, 1, maxPage));
+            }
+          }}
+          onKeyPress={handleKeyPress}
+          onBlur={() => {
+            setPage(inputValue);
+          }}
+        >
+          <NumberInputField
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-tool/pagination-tool.tsx:86]: We can't erase the number completely to set a new number, we have to set the new number and then delete the precedent one. Example: if I'm on page one and I want to type 2 to go to page 2, I need to first type number 2, so for a moment I have the number 12 in there, and then delete the 1

File: typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    setPagesAndItemsPerPages(maxPages, itemsPerPage);
+  }
+
+  return (
+    <Grid
+      pos="fixed"
+      bottom="0px"
+      w="100%"
+      h="44px"
+      zIndex={2}
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx:50]: Why do we need this?

File: typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  if (maxPages && currentPage > maxPages) {
+    setPagesAndItemsPerPages(maxPages, itemsPerPage);
+  }
+
+  return (
+    <Grid
+      pos="fixed"
+      bottom="0px"
+      w="100%"
+      h="44px"
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx:49]: Unless necessary, try to use Chakra values instead of directly pixels (same thing goes for everywhere you're using pixel values)

File: typescript/web/src/components/toggle-button-group/toggle-button.tsx [hunk 9]
```diff
@@ -30,13 +30,13 @@ export const ToggleButton = (props: ToggleButtonProps) => {
...
         <input {...inputProps} aria-labelledby={id} />
-        <IconButton
+        <Button
           as="div"
           id={id}
           color={useColorModeValue("gray.600", "whiteAlpha.700")}
           _checked={{
             color: useColorModeValue("inherit", "whiteAlpha.900"),
-            bg: useColorModeValue("gray.200", "whiteAlpha.300"),
+            bg: useColorModeValue("gray.100", "whiteAlpha.300"),
```
Comment (nacho-carnicero) [inline typescript/web/src/components/toggle-button-group/toggle-button.tsx:39]: Why this change?

File: typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        >
+          {Object.keys(itemsPerPageStrToNum).map((itemsPerPageOption) => (
+            <ToggleButton
+              size="sm"
+              key={`imagesPerPageButton-${itemsPerPageOption}`}
+              value={itemsPerPageOption}
+              description={`${itemsPerPageOption} images per page`}
+              pointerEvents="initial"
+              borderColor={mode("gray.100", "whiteAlpha.300")}
+              _checked={{
```
Comment (nacho-carnicero) [inline typescript/web/src/components/pagination-toolbar/pagination-toolbar.tsx:74]: This is the same as the default values set in the ToggleButton component no? I would say you don't need to pass it

**Comment by nacho-carnicero** [comment]
@valerian I did the review you asked for, apart from some minor stuff it's looking good so far, good job!

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx (hunk 3 above)
Comment (valerian) [reply typescript/web/src/components/pagination-tool/pagination-tool.tsx:93]: I have taken this code from the component used for navigating between image in the labelling view, it allows for scaling the number input field to fit the same number of digits as the maximum page number

**Review by bbenoist — CHANGES_REQUESTED** [review]
_:warning: I've not reviewed the Apollo cache parts in details yet._

File: typescript/web/src/components/pagination-tool/pagination-tool.tsx (hunk 3 above)
Comment (valerian) [reply typescript/web/src/components/pagination-tool/pagination-tool.tsx:93]: @crubier do you have an answer to that? It seems that you originally coded that part on the [image navigation tool](https://github.com/labelflow/labelflow/blob/752c7389a11e32af6c4086facc2a97e4c0c4c858/typescript/web-app/src/components/image-navigation-tool-bar/image-navigation-tool.tsx), which I partially copied

Comment (vincent-lecrubier-skydio) [reply typescript/web/src/components/pagination-tool/pagination-tool.tsx:93]: This number only depends on the font family used. We define the font in CSS so there is no reason for this number to change. It does not depend on zoom or font scaling since it is expressed in rem which also scale with zoom or font scaling, canceling each other. 

I found this number by carefully trying various values in storybook, and finding the optimal value that: 

- is not too wide, to avoid having an overly large margin in the box when the number is made of narrow digits (worst case is 11111)
- is not too narrow, to avoid cropping/overflowing the box when the number is made of wide digits (worst case is 88888 or 66666 i think)

The value is precise and 0.6 is too large and 0.5 is too small. 

I don’t think there is an easy way to test that, and anyway you’d be testing the test engine font rendering more than your own code. Also this value is not likely to change often so there is no reason to spend energy testing it really.

Comment (vincent-lecrubier-skydio) [reply typescript/web/src/components/pagination-tool/pagination-tool.tsx:93]: There are other methods of fitting an input box size to its content but 

- it’s better to only depend on the number of digits and not on the actual exact width of the number, to avoid constantly changing the width of the box every time the number change, which would be bad UX  (here we only change the width when we go from 9 to 10, … , 9999 to 10000 etc..)
- the methods to fit an input to its content often involve complex measurement process that involve DOM operations and CSS which are too costly for the need

File: typescript/web/src/components/dataset-images-list/image-card.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+              {name}
+            </Heading>
+            <IconButton
+              icon={<TrashIcon />}
+              aria-label="delete image"
+              isRound
+              size="sm"
+              onClick={(e) => {
+                e.preventDefault();
+                e.stopPropagation();
```
[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/dataset-images-list/image-card.tsx:63]: good question, it comes from this https://github.com/labelflow/labelflow/commit/4d595801c18842722a83191deae3827cd09aefca
@nacho-carnicero

File: typescript/web/src/components/dataset-images-list/images-list.tsx [hunk 12]
```diff
@@ -1,149 +1,134 @@
...
 
-const TrashIcon = chakra(HiTrash);
+const imagesPerPageOptions = [50, 200, 1000];
 
 export const ImagesList = ({
   datasetSlug,
   workspaceSlug,
-  images,
+  datasetId,
+  imagesTotalCount,
```
[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/dataset-images-list/images-list.tsx:27]: It was to be consistent with the GraphQL name, coming from
```
dataset {
    imagesAggregates {
        totalCount
    }
```

**Review by bbenoist — APPROVED** [review]

File: typescript/web/src/components/pagination/page-input.context.tsx [hunk 13]
```diff
@@ -48,7 +48,7 @@ const useHandleKeyDown = (reset: () => void) => {
   const { setPage } = usePagination();
   return (event: KeyboardEvent<HTMLInputElement>) => {
     if (event.key === "Enter") {
-      const newPage = parseInputValue(event.currentTarget.value);
+      const newPage = parseInputValue((event.target as HTMLInputElement).value);
```
Comment (bbenoist) [inline typescript/web/src/components/pagination/page-input.context.tsx:51]: @valerian Sorry. I thought `currentTarget` was defined in this case 😢

File: typescript/web/src/connectors/apollo-client/cache-config.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  existing,
+  options: FieldFunctionOptions<QueryImagesArgs, Record<string, any>>
+) => {
+  // A read function should always return undefined if existing is
+  // undefined. Returning undefined signals that the field is
+  // missing from the cache, which instructs Apollo Client to
+  // fetch its value from your GraphQL server.
+  if (existing === undefined) {
+    return undefined;
+  }
```
Comment (bbenoist) [inline typescript/web/src/connectors/apollo-client/cache-config.ts:24]:
```suggestion
  if (existing === undefined) return undefined;
```

File: typescript/web/src/connectors/apollo-client/cache-config.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  options: FieldFunctionOptions<QueryImagesArgs, Record<string, any>>
+) => {
+  // Slicing is necessary because the existing data is
+  // immutable, and frozen in development.
+  const merged = existing
+    ? {
+        images: existing.images.slice(0),
+        totalCount: existing.totalCount,
+      }
+    : { images: [], totalCount: undefined };
```
Comment (bbenoist) [inline typescript/web/src/connectors/apollo-client/cache-config.ts:60]:
```suggestion
  // Spreading the images array is necessary because the existing data is
  // immutable, and frozen in development
  const merged = {
    images: [...(existing?.images ?? []),
    totalCount: existing?.totalCount
  }
```

File: typescript/web/src/connectors/apollo-client/cache-config.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+  const offset = options.args?.skip || 0;
+  const limit = options.args?.first || undefined;
+  for (let i = 0; i < incoming.length; i += 1) {
+    merged.images[offset + i] = incoming[i];
+  }
```
Comment (bbenoist) [inline typescript/web/src/connectors/apollo-client/cache-config.ts:65]: _No really a change-request but rather a recommendation._
For such cases, I'd use [immutability-helper](https://www.npmjs.com/package/immutability-helper) which can, in a single operation, take care of making a deep copy (_but without being subject to object properties changes_) and setting only the updated values in the array. It's been made exactly to cover such cases (_it was created by the React team to help updating deep states_)

File: typescript/web/src/connectors/apollo-client/cache-config.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        )
+      : existing.images;
+  if (
+    result.length === 0 ||
+    (result as (Image | undefined)[]).includes(undefined) ||
+    // Invalidate cache if returned result is smaller than the limit
+    // but does not match the expected total count of images in the
+    // dataset
+    (((limit && result.length < limit) || !limit) &&
+      (!existing.totalCount || existing.totalCount !== offset + result.length))
```
Comment (bbenoist) [inline typescript/web/src/connectors/apollo-client/cache-config.ts:41]: That's what I call a big condition 😅 Can we split it into multiple well-named parts to reduce it's reading complexity?

**Review by bbenoist — CHANGES_REQUESTED** [review]

File: typescript/web/src/components/pagination/pagination.context.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+const usePaginationState = ({
+  page: defaultPage = 1,
+  perPageOptions = DEFAULT_PER_PAGE_OPTIONS,
+  perPage: defaultPerPage = undefined,
+  itemCount,
+}: Omit<PaginationProps, "children">): PaginationState => {
+  const [pagination, setPagination] = useReducer(paginationReducer, {
+    page: defaultPage,
+    perPage: defaultPerPage ?? perPageOptions[0],
```
[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/pagination/pagination.context.tsx:79]: I didn't know this was possible in TS/JS !

File: typescript/web/src/components/pagination/pagination.context.tsx [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+export const usePagination = () => useContext(PaginationContext);
+
+const paginationReducer = (
+  prevState: PaginationPayload,
+  nextState: Partial<PaginationPayload>
+): PaginationPayload => ({ ...prevState, ...nextState });
+
+export type PaginationProps = Pick<PaginationState, "itemCount"> &
+  Pick<PaginationState, "perPageOptions"> &
```
[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/pagination/pagination.context.tsx:33]: My bad, tried `,`, forgot about `|`

File: typescript/web/src/connectors/apollo-client/cache-config.ts (hunk 14 above)
Comment (valerian) [reply typescript/web/src/connectors/apollo-client/cache-config.ts:24]: I thought we enforced the use of curly brackets ?

Comment (bbenoist) [reply typescript/web/src/connectors/apollo-client/cache-config.ts:24]: I do this (**only for simple one-liner early returns**) since my arrival and never heard it was forbidden 😉

File: typescript/web/src/connectors/apollo-client/cache-config.ts (hunk 16 above)
Comment (valerian) [reply typescript/web/src/connectors/apollo-client/cache-config.ts:65]: I don't think immutability-helper would be of much help in that situation, in which I need to offset the indexes beforehand anyway.

File: typescript/web/src/components/dataset-images-list/paginated-images-query.ts [hunk 20]
```diff
@@ -16,12 +22,16 @@ export const flushPaginatedImagesCache = async (
   apolloClient: ApolloClient<object>,
   datasetId: string
 ) => {
-  // The custom variable "flush" is only expected by our custom
-  // apollo cache merge function, and will invalidate the cache
-  await apolloClient.writeQuery({
-    query: paginatedImagesQuery,
-    data: { images: [] },
-    variables: { where: { datasetId }, flush: true },
+  const storeFieldName = `images:${JSON.stringify({ where: { datasetId } })}`;
```
[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/dataset-images-list/paginated-images-query.ts:25]: Please see this line: https://github.com/labelflow/labelflow/blob/d040e16d76f32a58804850a1f346685b6c16bda3/typescript/web/src/connectors/apollo-client/cache-config.ts#L75

And that:
https://github.com/apollographql/apollo-client/blob/main/src/cache/inmemory/object-canon.ts

And in the current situation there is always only one property, so no risk of sort order issues.

[the remark this answers is not in the document]
Comment (valerian) [reply typescript/web/src/components/dataset-images-list/paginated-images-query.ts:25]: By the way I just realized I can use `canonicalStringify` from `object-canon.ts` instead of `JSON.stringify`, that would be better here.

**Review by bbenoist — APPROVED** [review]
